Genetic algorithm to fine tune strategy parameters

arphen opened the pull request:
I don't think it's in a state to be merged, but i'd like to have a discussion

Commit:
Added a genetic algorithm to train the parameters of the default strategy. To run:

`docker-compose up`
`docker-compose exec server bash`
`python zen/main.py <days> gdax.BTC-EUR`

Commit:
Added a genetic algorithm to train the parameters of the default strategy. To run:

`docker-compose up`
`docker-compose exec server bash`
`python zen/main.py  gdax.BTC-EUR <days>`

Commit:
Fixed a problem with pct being ill defined

Commit:
Improvements to the genetic algorithms persistance, hall of fame added as well as geneaology.

Commit:
Improvements to the genetic algorithms persistance, hall of fame added as well as geneaology.

Commit:
Parameter tuning

Commit:
Don't have sex with yourself

Commit:
Drawing adjustments

Commit:
Improvements to metrics validation training and logging

Commit:
Increase hall of fame size

Commit:
Use custom metric

Commit:
Params to conf py

Commit:
Merge branch 'master' of https://github.com/carlos8f/zenbot

Commit:
Refactor genetic algorithm

Commit:
Refactor genetic algorithm

Commit:
Refactor genetic algorithm

Commit:
Refactor genetic algorithm

miro-ka committed:
initial implementation of deep-learning strategy

miro-ka committed:
Merge github.com:carlos8f/zenbot

Commit:
Reduce penalty

miro-ka committed:
partial commit of q-learning strategy

Commit:
fuck you git

Commit:
somebody keeps erasing my changes.

miro-ka committed:
synching

Commit:
Now we tell you your browser sucks in your native tongue.

Commit:
Formatting output

Commit:
Fuzz training across products

Commit:
Fuzz training across products

Commit:
Remove debug, sort elites

Commit:
adjust slippage

Commit:
sort elites

miro-ka committed:
started to integrate talib

Commit:
One little whitespace gets its very own commit! Oh, life is so erratic!

Commit:
It only compiles every 2 tries... good luck.

Commit:
added some filthy stuff

Commit:
aslkdlj

Commit:
Merge branch 'master' of https://github.com/carlos8f/zenbot

# Conflicts:
#	package-lock.json

XavierGeerinck commented:
Looking nice @arpheno! Do you got any results of it?

Commit:
Issues with typing

Commit:
When you make a pull request, be sure that the damn thing actually works

arphen commented:
I have several tuned parameter constellations that for simulations are extremely profitable. One problem of simulations is that the slippage in real markets is much bigger than what is assumed in the simulator, so high frequency strategies really don't work out well.

I have 60 individuals across 60 days in three partitions running on my box right now, since yesterday night and waiting for results. These are trained across different currencies and different exchanges with a max-worst-case objective. 

Genetic algorithms are extremely good at learning crap, so if the data has quirks you get results over 12000% profit.

Commit:
Adjust log messages

XavierGeerinck commented:
Awesome! Would you be able to provide some documentation afterwards too? :)

arphen commented:
#298

Commit:
Readd minperiods

miro-ka committed:
adding AD and ADS inticators

miro-ka committed:
fixing strategy template bug

Commit:
hey, what's that over there?!

gelotus committed:
Merge remote-tracking branch 'genetics/master'

merge

gelotus committed:
Merge branch 'master' of https://github.com/carlos8f/zenbot

upstream

gelotus committed:
Merge branch 'master' of https://github.com/carlos8f/zenbot

gelotus committed:
Merge branch 'master' of https://github.com/carlos8f/zenbot

gelotus committed:
Merge remote-tracking branch 'deep/master'

gelotus committed:
Sync

gelotus committed:
Merge branch 'master' of https://github.com/carlos8f/zenbot

DeviaVir commented:
This is great and impressive work, thank you for contributing! I personally think this is much better than the current backtesting script. This definitely needs some work (variable selectors and perhaps separate docker comes to mind) but seeing it work as-is right now is great.

DeviaVir commented on Dockerfile:
```diff
@@ -1,6 +1,97 @@
 WORKDIR /app
+RUN /usr/local/bin/npm install
+RUN npm update
+
```
I would suggest to just create your own docker.

Also, to run this, I had to run `mkdir -p logs/hof`, so perhaps a `RUN mkdir -p zen/logs/hof` here around the end?

ivanmladek replied:
I get a weird error related to npm in docker. When you guys were building your docker, did you stumble upon this? 
server_1        | module.js:471
server_1        |     throw err;
server_1        |     ^
server_1        | 
server_1        | Error: Cannot find module 'semver'

nodevar replied:
If you are using arphenos dockerfile you need to copy at least package.json into you workdir before the npm install command.

or just copy the app into the working dir: 
COPY . /app

DeviaVir commented on fabfile.py:
```diff
@@ -0,0 +1,37 @@
+from fabric.api import run, cd,local
+
+
+
```
nit: one newline too much

DeviaVir commented on zen/.idea/inspectionProfiles/profiles_settings.xml:
```diff
@@ -0,0 +1,7 @@
+    <option name="USE_PROJECT_PROFILE" value="false" />
+    <version value="1.0" />
+  </settings>
+</component>
```
this is an IDE file, I would recommend you add `.idea` to your own global gitignore file

DeviaVir commented on zen/.idea/misc.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="ProjectRootManager" version="2" project-jdk-name="Python 2.7.12 (/usr/bin/python2.7)" project-jdk-type="Python SDK" />
+</project>
```
this is an IDE file, I would recommend you add `.idea` to your own global gitignore file

DeviaVir commented on zen/.idea/modules.xml:
```diff
@@ -0,0 +1,8 @@
+      <module fileurl="file://$PROJECT_DIR$/.idea/zen.iml" filepath="$PROJECT_DIR$/.idea/zen.iml" />
+    </modules>
+  </component>
+</project>
```
this is an IDE file, I would recommend you add `.idea` to your own global gitignore file

DeviaVir commented on zen/.idea/zen.iml:
```diff
@@ -0,0 +1,15 @@
+    <option name="projectConfiguration" value="py.test" />
+    <option name="PROJECT_TEST_RUNNER" value="py.test" />
+  </component>
+</module>
```
this is an IDE file, I would recommend you add `.idea` to your own global gitignore file

DeviaVir commented on zen/.idea/workspace.xml:
```diff
@@ -0,0 +1,736 @@
+      </provider>
+    </entry>
+  </component>
+</project>
```
this is an IDE file, I would recommend you add `.idea` to your own global gitignore file

DeviaVir commented on zen/conf.py:
```diff
@@ -0,0 +1,13 @@
+
+
+
+
```
nit: too many newlines

DeviaVir commented on zen/evaluation.py:
```diff
@@ -0,0 +1,143 @@
+    >>>fuzz_product('USD')"""
+    selectors = {
+        'BTC-CUR': ['gdax.BTC-USD', 'gdax.BTC-EUR', 'gdax.BTC-GBP'],
+        'ETH-BTC': ['poloniex.ETH-BTC', 'gdax.ETH-BTC']
```
I understand why this is a constant now (for demo purposes) but if we plan on merging this in at some point, we should try and use the `list-selectors` from zenbot, perhaps?

DeviaVir commented on fabfile.py:
```diff
@@ -0,0 +1,37 @@
+        cmd = '/app/zenbot.sh backfill {instrument} --days {days}'.format(days=TOTAL_DAYS, instrument=instrument)
+        remote(cmd,'backfill_'+instrument)
+def backfill_local(TOTAL_DAYS):
+    products = ['gdax.BTC-EUR','gdax.BTC-USD','gdax.BTC-GBP']+['gdax.ETH-BTC','poloniex.ETH-BTC']
```
I understand why this is a constant now (for demo purposes) but if we plan on merging this in at some point, we should try and use the `list-selectors` from zenbot, perhaps?

Commit:
Merge branch 'master' of https://github.com/carlos8f/zenbot

Commit:
minor changes

gelotus committed:
Sync

gelotus committed:
fix

gelotus committed:
Merge remote-tracking branch 'genetics/master'

nao0811ta commented:
question.

It have never been evaluated during few sleeping hours.
```
[ec2-user@ip-10-166-11-188 zen]$ python -m scoop main.py ETH-BTC 14 10 macd
[2017-06-20 15:51:58,504] launcher  INFO    SCOOP 0.7 1.1 on linux using Python 3.6.1 (default, Jun 20 2017, 15:15:55) [GCC 4.8.3 20140911 (Red Hat 4.8.3
-9)], API: 1013
[2017-06-20 15:51:58,504] launcher  INFO    Deploying 2 worker(s) over 1 host(s).
[2017-06-20 15:51:58,505] launcher  INFO    Worker distribution:
[2017-06-20 15:51:58,505] launcher  INFO       127.0.0.1:       1 + origin
MAKE SURE YOU RUN fab backfill_local:<days>
otherwise it's all crap
Starting evolution....
Evaluating 10 individuals over 14 days in 2 partitions.
Mating function is <function cxTwoPoint at 0x7fc3fea96268>
Mutating function is functools.partial(<function mutGaussian at 0x7fc3fbee88c8>, mu=0, sigma=20, indpb=0.7)
Generating an initial valid population, this may take a while...
Summer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxxxxxxxxSummer is here, need to evaluate 10 individuals
          |
xxxx-xxxxxSummer is here, need to evaluate 10 individuals
<abbr asame output>
```

nao0811ta commented:
FileNotFoundError: [Errno 2] No such file or directory: '/app/zenbot.sh'

I wish that zenbot.sh is relative path
and saimulation saving path also([#317](https://github.com/carlos8f/zenbot/issues/317)).

arphen commented:
I screwed up something in master, my bad. I'm fixing it right now.

On Wed, Jun 21, 2017 at 6:16 AM Takatoshi Nao <notifications@github.com>
wrote:

> FileNotFoundError: [Errno 2] No such file or directory: '/app/zenbot.sh'
>
> I wish that zenbot.sh is relative path
> and saimulation saving path also.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/carlos8f/zenbot/pull/299#issuecomment-309958343>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABBzGiQ7oLsmDbASCakVuMTaHmWr7_1Qks5sGJkugaJpZM4N8epK>
> .
>

nao0811ta commented:
It takes so long time. So 
How about save and restart function?

Commit:
repair master

Commit:
repair master

nao0811ta commented:
```
  File "main.py", line 41, in <module>
    res = main(INSTRUMENT, TOTAL_DAYS, popsize, strategy)
  File "main.py", line 24, in main
    res = evolve(evaluate, Individual, popsize)
  File "/home/ec2-user/workspace/zen4-polo/zen/evolution/__init__.py", line 27, in evolve
    return algorithm(toolbox, stats, history, hof,popsize)
  File "/home/ec2-user/workspace/zen4-polo/zen/evolution/core.py", line 16, in algorithm
    evaluate(would_be, toolbox)
  File "/home/ec2-user/workspace/zen4-polo/zen/evolution/core.py", line 40, in evaluate
    for ind, fit in zip(invalid_ind, fitnesses):
  File "/home/ec2-user/.pyenv/versions/zen-polo/lib/python3.6/site-packages/scoop/futures.py", line 102, in _mapGenerator
    for future in _waitAll(*futures):
  File "/home/ec2-user/.pyenv/versions/zen-polo/lib/python3.6/site-packages/scoop/futures.py", line 358, in _waitAll
    for f in _waitAny(future):
  File "/home/ec2-user/.pyenv/versions/zen-polo/lib/python3.6/site-packages/scoop/futures.py", line 335, in _waitAny
    raise childFuture.exceptionValue
IndexError: Cannot choose from an empty sequence
```

Commit:
repair master

arphen commented:
Yep sorry guys for the inconvenience. I fucked up the master branch and learned yet again not to commit to master directly. Multiprocessing is more exciting than it should be.

Things are working again.

gelotus committed:
Merge remote-tracking branch 'genetics/master'

gelotus committed:
Path fix

nao0811ta commented:
@arpheno 
thanks!! it works again.

arphen commented:
Here's some results:

```
user@genericbox:~/zenbot$ tail -f ETH*30*.out
==> ETH-BTC_macd_50_30_2017-06-22-18-04.out <==
...

Generation 156 {'avg': -4.3447320162938903, 'std': 23.51224732056831, 'min': -39.009999999999998, 'max': 19.99943054239105, 'len': 30}
It's breeding season, we're expecting new members of the tribe...
10 children have been born.
Radiation and toxic waste are causing mutations in the population...
8 individuals have mutated.
Summer is here, evaluating our new arrivals...
                  |
........x........
==> ETH-BTC_rsi_50_30_2017-06-22-18-06.out <==
...

Generation 191 {'avg': -13.496419041828347, 'std': 13.552027774607932, 'min': -31.670000000000002, 'max': 1.0922793968148585, 'len': 30}
It's breeding season, we're expecting new members of the tribe...
10 children have been born.
Radiation and toxic waste are causing mutations in the population...
9 individuals have mutated.
Summer is here, evaluating our new arrivals...
                   |
................
==> ETH-BTC_sar_50_30_2017-06-22-13-42.out <==
...

Generation 366 {'avg': -3.786960662691008, 'std': 9.7398293425581581, 'min': -30.240098048429321, 'max': 9.0500729772017543, 'len': 30}
It's breeding season, we're expecting new members of the tribe...
10 children have been born.
Radiation and toxic waste are causing mutations in the population...
11 individuals have mutated.
Summer is here, evaluating our new arrivals...
                     |
...................
==> ETH-BTC_trend_ema_50_30_2017-06-22-18-06.out <==
...

Generation 162 {'avg': -15.808718271857389, 'std': 21.800960754046944, 'min': -100.69314718055995, 'max': 4.2826980234809708, 'len': 30}
It's breeding season, we're expecting new members of the tribe...
10 children have been born.
Radiation and toxic waste are causing mutations in the population...
9 individuals have mutated.
Summer is here, evaluating our new arrivals...
                   |

```

gelotus committed:
In population is low len is 0. Is better a fixed value

gelotus committed:
Working dir variable in conf

gelotus committed:
Merge remote-tracking branch 'lead/master'

gelotus committed:
Ignore pyc

gelotus committed:
Merge remote-tracking branch 'lead/master'

gelotus committed:
Replace colorterm with blessings

gelotus committed:
Remove escape characters from hall of fame file

gelotus committed:
Merge remote-tracking branch 'lead/master'

gelotus committed:
More envinronment agnostic and new syntax for backfill

fab backfill_local:"<selector>,<days>"

gelotus committed:
Implement auto backfill timer 

For activating append backfill at te end of commandline. Interval can be set in conf

gelotus committed:
Add placeholders

arphen committed:
Merge pull request #2 from gelotus/master

Implement working dir as variable editable in conf.

nao0811ta commented:
Thank you all
DQN Awesome..

Do you know why happen like below?
```
python -m scoop main.py ETH-BTC 10 30 macd backfill
[2017-06-28 15:51:12,306] launcher  INFO    SCOOP 0.7 1.1 on linux using Python 3.6.1 (default, Jun 20 2017, 15:15:55) [GCC 4.8.3 20140911 (Red Hat 4.8.3-9)], API: 1013
[2017-06-28 15:51:12,306] launcher  INFO    Deploying 2 worker(s) over 1 host(s).
[2017-06-28 15:51:12,306] launcher  INFO    Worker distribution:
[2017-06-28 15:51:12,306] launcher  INFO       127.0.0.1:       1 + origin 
Auto backfilling is ON
Executing inital backfilling, please wait...
Traceback (most recent call last):
  File "/home/ec2-user/.pyenv/versions/3.6.1/envs/zen-polo/lib/python3.6/site-packages/fabric/main.py", line 672, in main
    docstring, callables, default = load_fabfile(fabfile)
  File "/home/ec2-user/.pyenv/versions/3.6.1/envs/zen-polo/lib/python3.6/site-packages/fabric/main.py", line 173, in load_fabfile
    imported = importer(os.path.splitext(fabfile)[0])
  File "/home/ec2-user/workspace/zen4-polo/fabfile.py", line 5, in <module>
    from zen import path, pyexc, selectors
  File "/home/ec2-user/workspace/zen4-polo/zen/__init__.py", line 1, in <module>
    from conf import path, pyexc, selectors
ImportError: cannot import name 'path'
```

arphen committed:
Revert "Implement working dir as variable editable in conf."

arphen committed:
Merge pull request #3 from arpheno/revert-2-master

Revert "Implement working dir as variable editable in conf."

Yroethiel commented:
Any suggestions why "docker-compose build" fails with this Dockerfile?

It breaks on Step 4/17:

`RUN apt-get update && apt-get install -y --no-install-recommends tcl tk && rm -rf /var/lib/apt/lists/*`

The error I get is:

`ERROR: Service 'server' failed to build: the command '/bin/sh apt-get update && apt-get install -y --no-install-recommends tcl tk && rm -rf /var/lib/apt/lists/*' returned a non-zero code: 1 `

Latest version of Debian Jessie, latest python, latest npm, latest Docker. Tried on ubuntu as well.

arphen commented:
It really shouldn't happen. Did you git pull the newest version? Somethings
not right with the last PR

On Wed, Jun 28, 2017 at 7:22 PM Yroethiel <notifications@github.com> wrote:

> Any suggestions why "docker-compose build" fails with this Dockerfile?
>
> It breaks on Step 4/17:
>
> RUN apt-get update && apt-get install -y --no-install-recommends tcl tk &&
> rm -rf /var/lib/apt/lists/*
>
> The error I get is:
>
> ERROR: Service 'server' failed to build: the command '/bin/sh apt-get
> update && apt-get install -y --no-install-recommends tcl tk && rm -rf
> /var/lib/apt/lists/*' returned a non-zero code: 1
>
> Latest version of Debian Jessie, latest python, latest npm, latest Docker.
> Tried on ubuntu as well.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/carlos8f/zenbot/pull/299#issuecomment-311729087>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABBzGkq0Dt6AXObaduIlwQ74M9LelwcMks5sIovPgaJpZM4N8epK>
> .
>

Yroethiel commented:
Yeah, has me confused as well. All cloned via git, using today's release. Same issue on the release from May 22.

sandroid10 commented:
@arpheno  i am getting this error inside docker ( i did backfill) 

Evaluating 10 individuals over 3 days in 2 partitions.
Mating function is <function cxTwoPoint at 0x7f9dc17ce400>
Mutating function is functools.partial(<function mutGaussian at 0x7f9dc01efae8>, mu=0, sigma=20, indpb=0.3)
Sampling an initial valid population, this may take a while...
Currently 0 valid individuals
          |
...x......Currently 5 valid individuals
          |
..........Traceback (most recent call last):
  File "/usr/local/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.6/site-packages/scoop/bootstrap/__main__.py", line 302, in <module>
    b.main()
  File "/usr/local/lib/python3.6/site-packages/scoop/bootstrap/__main__.py", line 92, in main
    self.run()
  File "/usr/local/lib/python3.6/site-packages/scoop/bootstrap/__main__.py", line 290, in run
    futures_startup()
  File "/usr/local/lib/python3.6/site-packages/scoop/bootstrap/__main__.py", line 271, in futures_startup
    run_name="__main__"
  File "/usr/local/lib/python3.6/site-packages/scoop/futures.py", line 64, in _startup
    result = _controller.switch(rootFuture, *args, **kargs)
  File "/usr/local/lib/python3.6/site-packages/scoop/_control.py", line 253, in runController
    raise future.exceptionValue
  File "/usr/local/lib/python3.6/site-packages/scoop/_control.py", line 127, in runFuture
    future.resultValue = future.callable(*future.args, **future.kargs)
  File "/usr/local/lib/python3.6/runpy.py", line 263, in run_path
    pkg_name=pkg_name, script_name=fname)
  File "/usr/local/lib/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/local/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "main.py", line 42, in <module>
    res = main(INSTRUMENT, TOTAL_DAYS, popsize, strategy)
  File "main.py", line 25, in main
    res = evolve(evaluate, Andividual, popsize)
  File "/app/zen/evolution/__init__.py", line 18, in evolve
    return algorithm(cls, popsize, futures.map, evaluate, select, breed, mutate, stats, history, hof)
  File "/app/zen/evolution/core.py", line 21, in algorithm
    log_stuff(g, history, hof, population, stats)
  File "/app/zen/evolution/utils.py", line 30, in log_stuff
    hof.persist()
  File "/app/zen/halloffame.py", line 27, in persist
    with open('logs/hof/{runid}.txt'.format(runid=runid), 'w') as f:
FileNotFoundError: [Errno 2] No such file or directory: 'logs/hof/3520.txt'
[2017-07-25 17:02:54,946] launcher  (127.0.0.1:37786) INFO    Root process is done.
[2017-07-25 17:02:54,946] launcher  (127.0.0.1:37786) INFO    Finished cleaning spawned subprocesses.

Ekliptor commented:
What's the difference between elites and specialists in your selection algorithm? I understand the goal of the algorithm is to find the most profitable strategy parameters. So what is the criteria for "profitable" for elites and specialists?

DeviaVir reviewed (changes requested):
@arpheno I would like to merge your PR, but there are still a few unhandled comments and things I would like to see changed.

The most important one, in my eyes, is that this should live in the `/scripts/` folder and not touch so many other files that have relatively little to do with the proposed additions.

evseevnn-zz commented:
@arpheno That work? how soon you wanna solve problem of PR?
May be somebody can help?

DeviaVir commented:
Superseded by https://github.com/carlos8f/zenbot/pull/598
